src/impl_/pyclass_init.rs

```diff
@@ -17,69 +19,70 @@ pub trait PyObjectInit<T>: Sized + Sealed {
+        kwargs: Option<&Bound<'_, PyDict>>,
     ) -> PyResult<*mut ffi::PyObject>;
 
     #[doc(hidden)]
     fn can_be_subclassed(&self) -> bool;
 }
 
-/// Initializer for Python native types, like `PyDict`.
-pub struct PyNativeTypeInitializer<T: PyTypeInfo>(pub PhantomData<T>);
+/// Initializer for Python native types, like [PyDict].
+pub struct PyNativeTypeInitializer<T: PyTypeInfo + PyClassBaseType>(pub PhantomData<T>);
 
-impl<T: PyTypeInfo> PyObjectInit<T> for PyNativeTypeInitializer<T> {
+impl<T: PyTypeInfo + PyClassBaseType> PyObjectInit<T> for PyNativeTypeInitializer<T> {
+    /// call `__new__` ([ffi::PyTypeObject::tp_new]) for the native base type.
+    /// This will allocate a new python object and initialize the part relating to the native base type.
     unsafe fn into_new_object(
         self,
         py: Python<'_>,
         subtype: *mut PyTypeObject,
+        args: &Bound<'_, PyTuple>,
+        kwargs: Option<&Bound<'_, PyDict>>,
     ) -> PyResult<*mut ffi::PyObject> {
         unsafe fn inner(
             py: Python<'_>,
-            type_object: *mut PyTypeObject,
+            native_base_type: *mut PyTypeObject,
             subtype: *mut PyTypeObject,
+            args: &Bound<'_, PyTuple>,
+            kwargs: Option<&Bound<'_, PyDict>>,
+            new_accepts_arguments: bool,
         ) -> PyResult<*mut ffi::PyObject> {
-            // HACK (due to FIXME below): PyBaseObject_Type's tp_new isn't happy with NULL arguments
-            let is_base_object = type_object == std::ptr::addr_of_mut!(ffi::PyBaseObject_Type);
-            let subtype_borrowed: Borrowed<'_, '_, PyType> = subtype
+            let native_base_type_borrowed: Borrowed<'_, '_, PyType> = native_base_type
                 .cast::<ffi::PyObject>()
                 .assume_borrowed_unchecked(py)
                 .downcast_unchecked();
+            let tp_new = native_base_type_borrowed
+                .get_slot(TP_NEW)
+                .unwrap_or(ffi::PyType_GenericNew);
 
-            if is_base_object {
-                let alloc = subtype_borrowed
-                    .get_slot(TP_ALLOC)
-                    .unwrap_or(ffi::PyType_GenericAlloc);
-
-                let obj = alloc(subtype, 0);
-                return if obj.is_null() {
-                    Err(PyErr::fetch(py))
-                } else {
-                    Ok(obj)
-                };
-            }
-
-            #[cfg(Py_LIMITED_API)]
-            unreachable!("subclassing native types is not possible with the `abi3` feature");
+            let obj = if new_accepts_arguments {
+                tp_new(
+                    subtype,
+                    args.as_ptr(),
+                    kwargs
+                        .map(|obj| obj.as_ptr())
+                        .unwrap_or(std::ptr::null_mut()),
+                )
+            } else {
+                let args = PyTuple::empty(py);
+                tp_new(subtype, args.as_ptr(), std::ptr::null_mut())
+            };
```

Comment:
using `tp_new` instead of a special case for `PyBaseObject_Type` is nice in that the purpose of this function is clearer (it calls `__new__`) but since extending `PyBaseObject_Type` is so common it could also be specialised like before.

Currently `PyBaseObject_Type::tp_new` does some extra work:
- checks for abstractmethods
- initializes `__dict__`. As tested in `managed_dict_initialized_as_expected` below, this dict is overwritten by `T::Dict::INIT` (nullptr) then initialized again on first access.

Since the first dict is leaked, the best thing to do may be to handle `PyBaseObject_Type` specially. Alternatively, handling of `dict` could be altered so that the one provided by `object.__new__` doesn't get overwritten? With the new approach for managing `__dict__` in python 3.12 these complications go away as the space for dict is managed by the python itself but because of backwards compatibility this won't simplify things in pyo3 for ages (unless specifically using `#[pyclass(dict)]` can be made to require 3.12+ ?). https://docs.python.org/3/c-api/typeobj.html#c.Py_TPFLAGS_MANAGED_DICT

Reply:
It seems that other base native types like `PyFloat_type` have custom implementations that may not initialize `__dict__` like `object.__new__` does. So that rules out the option to not overwrite with `T::Dict::INIT` because in some cases it will be set and others not. So `PyBaseObject_type` needs some special handling unless the `tp_dictoffset` approach to handling `__dict__` is removed.